### PR TITLE
34 view query op

### PR DESCRIPTION
**Pull request by ricellis**
*What*
Added support for querying a view

*How*
Added `QueryViewOperation` accepting properties for optional `queryViewCompletionHandler` and `rowHandler` for processing view results as well as properties for the request query options.

*Testing*
New test cases for querying a view and validation of operation.

reviewer @rhyshort 
reviewer @brynh

### CODE REVIEW

File: Source/QueryViewOperation.swift [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+     Include the full document content in the view results.
+     
+     Optional: default is `false`, do not include the entire document contents.
+     */
+    public var includeDocs: Bool? = nil
+    
+    /**
+     Use the reduce function for the view.
+     
+     Optional: default is `true`, use the reduce function if there is one defined in the view.
```
Comment (brynh) [inline Source/QueryViewOperation.swift]: Seems like a good solution and reflects what actually happens.

File: Source/QueryViewOperation.swift [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+     - Warning: Only valid if `reduce` and `group` are both true
+     */
+    public var groupLevel: Int? = nil
+    
+    /**
+     Acceptable values for the `stale` property.
+     */
+    public enum Stale: CustomStringConvertible {
+         /// Do not allow stale views.
+        case No
```
Comment (brynh) [inline Source/QueryViewOperation.swift]: I was tending to agree with @rhyshort prior to reading the issue @ricellis linked to, but having read that, I'm inclined to agree that leaving it like this makes things somewhat clearer if `QueryViewOperation`'s are reused, which is a valid use-case but one which doesn't immediately spring to mind.

File: Source/QueryViewOperation.swift [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+    
+    /**
+     Sets a completion handler to run when the operation completes.
+     
+     - parameter error: - ErrorProtocol instance with information about an error executing the operation
+     */
+    public var queryViewCompletionHandler: ((error: ErrorProtocol?) -> Void)?
+    
+    /**
+     Sets a handler to run for each row retrived by the view.
```
Comment (brynh) [inline Source/QueryViewOperation.swift]: Typo `retrived` -> `retrieved`.

File: Source/QueryViewOperation.swift (hunk 1 above)
Comment (rhyshort) [inline Source/QueryViewOperation.swift]: Yeah ditto. I think that's what we should use.

File: Tests/QueryViewTests.swift [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+                                                                            "key" : "Alabama peanut chicken",
+                                                                            "value" : [
+                                                                                          "Alabama peanut chicken"
+                                                                            ]
+                                                            ]
+                    ],
+                                                           "total_rows" : 2667
+                    ], statusCode: 200, headers: [:])
+        })
+
```
Comment (brynh) [inline Tests/QueryViewTests.swift]: Remove multiple blank lines.

File: Tests/QueryViewTests.swift [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+        view.descending = true
+        view.endKey = "endkey"
+        view.includeDocs = true
+        view.inclusiveEnd = true
+        view.limit = 4
+        view.skip = 0
+        view.stale = .Ok
+        view.startKey = "startKey"
+        view.databaseName = self.dbName
+
```
Comment (brynh) [inline Tests/QueryViewTests.swift]: Remove one of the blank lines.

File: Tests/QueryViewTests.swift [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+        view.descending = true
+        view.endKey = ["endkey","endkey2"]
+        view.includeDocs = true
+        view.inclusiveEnd = true
+        view.limit = 4
+        view.skip = 0
+        view.stale = .Ok
+        view.startKey = ["startKey","startKey2"]
+        view.databaseName = self.dbName
+
```
Comment (brynh) [inline Tests/QueryViewTests.swift]: Remove one of the blank lines.

File: Tests/QueryViewTests.swift [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+                                                    NSURLQueryItem(name: "inclusive_end", value: "true"),
+                                                    NSURLQueryItem(name: "limit", value: "4"),
+                                                    NSURLQueryItem(name: "skip", value: "0"),
+                                                    NSURLQueryItem(name: "stale", value: "ok"),
+                                                    NSURLQueryItem(name: "startkey", value: "[\"startKey\",\"startKey2\"]")
+            
+        ]
+        
+        XCTAssert(expectedQueryItems.isEquivalent(to: view.queryItems))
+
```
Comment (brynh) [inline Tests/QueryViewTests.swift]: Remove blank lines.

File: Source/QueryViewOperation.swift (hunk 2 above)
Comment (rhyshort) [inline Source/QueryViewOperation.swift]: I'm pretty sure we can't reuse NSOperation classes see this SO question http://stackoverflow.com/questions/16844806/nsoperationqueues-addoperation-operation-is-finished-and-cannot-be-enqueued

File: Tests/QueryViewTests.swift [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+        let expectedQueryItems = [ NSURLQueryItem(name: "descending", value: "true"),
+                                                    NSURLQueryItem(name: "endkey", value: "\"endkey\""),
+                                                    NSURLQueryItem(name: "include_docs", value: "true"),
+                                                    NSURLQueryItem(name: "inclusive_end", value: "true"),
+                                                    NSURLQueryItem(name: "limit", value: "4"),
+                                                    NSURLQueryItem(name: "skip", value: "0"),
+                                                    NSURLQueryItem(name: "stale", value: "ok"),
+                                                    NSURLQueryItem(name: "startkey", value: "\"startKey\"")
+            
+                                                      ]
```
Comment (brynh) [inline Tests/QueryViewTests.swift]: The positioning of the `]` in all these tests is inconsistent.

File: Tests/QueryViewTests.swift [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+        view.databaseName = self.dbName
+
+        
+        XCTAssert(view.validate())
+        XCTAssertEqual("GET",view.httpMethod)
+        XCTAssertNil(view.httpRequestBody)
+        XCTAssertEqual("/\(self.dbName!)/_design/ddoc/_view/view1",view.httpPath)
+        
+        let expectedQueryItems = [ NSURLQueryItem(name: "descending", value: "true"),
+                                                    NSURLQueryItem(name: "endkey", value: "\"endkey\""),
```
Comment (brynh) [inline Tests/QueryViewTests.swift]: Align these items with the first.

File: Tests/QueryViewTests.swift [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+    
+    func testViewGeneratesCorrectRequestForReduces() {
+        let view = QueryViewOperation()
+        view.designDoc = "ddoc"
+        view.viewName = "view1"
+        view.group = true
+        view.groupLevel = 3
+        view.reduce = true
+        view.databaseName = self.dbName
+
```
Comment (brynh) [inline Tests/QueryViewTests.swift]: There are loads of cases where there are multiple consecutive blank lines. I'd prefer these all to be max 1 blank, but I won't continue to comment them individually.

File: Tests/QueryViewTests.swift [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+            XCTAssertNil(error)
+        }
+        
+        self.client?[self.dbName!].add(operation: view)
+        self.waitForExpectations(withTimeout: 10.0, handler: nil)
+    }
+    
+    func testOperationValidationMissingDDoc() {
+        let view = QueryViewOperation()
+                view.databaseName = "dbname"
```
Comment (brynh) [inline Tests/QueryViewTests.swift]: Indentation

File: Tests/QueryViewTests.swift [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+        XCTAssert(view.validate())
+    }
+    
+    func testOperationValidationReduceWithoutReduceGroup(){
+        let view = QueryViewOperation()
+        view.databaseName = "dbname"
+        view.designDoc = "ddoc"
+        view.viewName = "view1"
+        view.group = true
+        XCTAssert(view.validate())
```
Comment (brynh) [inline Tests/QueryViewTests.swift]: Shouldn't this be  `XCTAssertFalse`

File: Source/QueryViewOperation.swift (hunk 1 above)
Comment (ricellis) [inline Source/QueryViewOperation.swift]: Fixed in dd7e4ff.

File: Source/QueryViewOperation.swift [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+     Use the reduce function for the view.
+     
+     Optional: default is `true`, use the reduce function if there is one defined in the view.
+     */
+    public var reduce: Bool? = nil
+    
+    /**
+     Group the results of a reduce based on their keys.
+     
+     Optional: default is `false`, result rows will not be grouped by key.
```
Comment (ricellis) [inline Source/QueryViewOperation.swift]: Fixed in dd7e4ff.

File: Source/QueryViewOperation.swift [hunk 14]
```diff
@@ -0,0 +1,382 @@
...
+    public var group: Bool? = nil
+    
+    /**
+     Control view aggregation of complex keys by setting the number of elements of the key array to use for grouping reduce results.
+     
+     For example, if the view emits complex keys of the form `["A", "B", "C"]` and is using a `_count` reduce function then setting `groupLevel` would have these effects:
+     * `view.groupLevel = 3` would produce rows of the form `{ "key" : ["A", "B", "C"], "value" : n }` where `n` is the count of documents with key `["A", "B", "C"]
+     * `view.groupLevel = 1` would produces rows of the form `{ "key" : ["A"], "value" : m }` where `m` is the aggregated count of all keys with "A" as the first element
+     
+     Optional: default is to group for all elements in the key
```
Comment (ricellis) [inline Source/QueryViewOperation.swift]: Fixed in dd7e4ff.

File: Source/QueryViewOperation.swift [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+            case .Ok: return "ok"
+            case .UpdateAfter: return "update_after"
+            }
+        }
+    }
+    
+    /**
+     Configures whether the view should be allowed to return stale results, that is allowing a view to return immediately rather than waiting for the view to build.
+     
+     Optional: default is `Stale.No`, do not return stale results.
```
Comment (ricellis) [inline Source/QueryViewOperation.swift]: Fixed in dd7e4ff.

File: Source/QueryViewOperation.swift (hunk 3 above)
Comment (ricellis) [inline Source/QueryViewOperation.swift]: Fixed in dd7e4ff.

File: Source/QueryViewOperation.swift (hunk 2 above)
Comment (ricellis) [inline Source/QueryViewOperation.swift]: I've removed `Stale.No` in e1ed685.

**Comment by ricellis** [comment]
Addressed test formatting issues in e700008. We will reformat everything at some point when a formatter is available anyway.

**Comment by rhyshort** [comment]
I think we need one more test, this time to make sure "update_after" comes out correctly other than that I think we should merge

**Comment by ricellis** [comment]
Added stale tests in 44de7f3 testing each state individually in isolation.

**Comment by brynh** [comment]
+1

**Pull request merged**